OTAnalytics/plugin_ui/customtkinter_gui/line_section.py

```diff
@@ -106,11 +155,290 @@ def delete(self, tag_or_id: str) -> None:
+        elif event_type == ESCAPE_KEY:
+            self._deselect_knob()
+
+    def _hover_knob(self, coordinate: tuple[int, int]) -> None:
+        closest_knob_index = self._get_closest_knob_index(coordinate=coordinate)
+        self._hovered_knob_index = closest_knob_index
+        if closest_knob_index is not None:
+            self._redraw_temporary_section(
+                highlighted_knob_index=closest_knob_index,
+                highlighted_knob_style=self._hovered_knob_style,
+            )
+        else:
+            self._redraw_temporary_section()
+
+    def _select_hovered_knob(self) -> None:
+        self._selected_knob_index = self._hovered_knob_index
+        if self._selected_knob_index is not None:
+            self._redraw_temporary_section(
+                highlighted_knob_index=self._selected_knob_index,
+                highlighted_knob_style=self._selected_knob_style,
+            )
+
+    def _deselect_knob(self) -> None:
+        self._selected_knob_index = None
+        self._temporary_coordinates = self._coordinates.copy()
+        self._redraw_temporary_section(
+            highlighted_knob_index=self._hovered_knob_index,
+            highlighted_knob_style=self._hovered_knob_style,
+        )
+
+    def _move_knob(self, coordinate: tuple[int, int]) -> None:
+        if self._selected_knob_index is not None:
+            self._update_temporary_coordinate(
+                index=self._selected_knob_index, coordinate=coordinate
+            )
+            self._redraw_temporary_section(
+                highlighted_knob_index=self._selected_knob_index,
+                highlighted_knob_style=self._selected_knob_style,
+            )
+
+    def _add_knob(self, coordinate: tuple[int, int]) -> None:
+        self._append_temporary_coordinate(coordinate=coordinate)
+        self._selected_knob_index = len(self._temporary_coordinates) - 1
+        self._redraw_temporary_section(
+            highlighted_knob_index=self._selected_knob_index,
+            highlighted_knob_style=self._selected_knob_style,
+        )
+
+    def _shift_selected_knob_forward(self) -> None:
+        i = self._selected_knob_index
+        if i is None or i >= len(self._temporary_coordinates) - 1:
+            return
+        self._swap_temporary_coordinates(index=i, index_other=i + 1)
+        self._selected_knob_index = i + 1
+        self._redraw_temporary_section(
+            highlighted_knob_index=self._selected_knob_index,
+            highlighted_knob_style=self._selected_knob_style,
+        )
+
+    def _shift_selected_knob_backward(self) -> None:
+        i = self._selected_knob_index
+        if i is None or i == 0:
+            return
+        self._swap_temporary_coordinates(index=i, index_other=i - 1)
+        self._selected_knob_index = i - 1
+        self._redraw_temporary_section(
+            highlighted_knob_index=self._selected_knob_index,
+            highlighted_knob_style=self._selected_knob_style,
+        )
```

Comment:
This functionality looks a bit confusing in UI. I assume, the index of the current selected coordinate is moved. Lets see how users react to this feature :-)

Reply:
Yes, from my point of view it is only temporary. But we need it as a tool to place new buttons at any position of the section. In the future we could, for example, display a "+" when hovering over a segment of a section to insert a new knob between two existing knobs.